PSBT RPC calls

niftynei opened the pull request:
These are the base RPC calls that we'll need for constructing a transaction iteratively. ~As a PR, this is built ontop of #3762. New commits start at https://github.com/ElementsProject/lightning/commit/569efe8ba540cadc453694d22a8f0ed281802851~

Added RPC methods:
- reserveinputs
- unreserveinputs
- signpsbt
- sendpsbt

niftynei commented:
Travis failures likely due to https://github.com/ElementsProject/libwally-core/pull/203

cdecker commented:
Rebased on top of `master` since #3762 was merged.

niftynei commented on wallet/wallet.h:
```diff
@@ -1246,6 +1246,20 @@ void add_unreleased_tx(struct wallet *w, struct unreleased_tx *utx);
 
+/* wallet_persist_utxo_reservation - Removes destructor
+ *
+ * Persists the reservation in the database (until a restart)
+ * instead of clearing the reservation when the utxo object
+ * is destroyed */
```
I think this is the code you're looking for https://github.com/ElementsProject/lightning/blob/master/wallet/wallet.c#L3723-L3758

niftynei replied:
I'm pulling out the 'reservation system' from #3418 right now, which will make reservations persistent/expirable.

cdecker replied:
That's an excellent idea, what timeout are you looking at for default? (pretty hesitant to add yet more arguments to commands, since 90% are simply unused)

cdecker replied:
Also, good work on finding the un-reserve on startup, I was getting lost looking for it :wink:

niftynei replied:
Ok so this is now available in #3789. I defaulted to 144 blocks, or 24 hours. You can disable it or choose a different number, but that'll revert to the current behavior of resetting the utxo reservation if the node is restarted.

One nice feature of the way we 'broadcast' txs is that if you send a tx via c-lightning it will mark it as 'spent' (really this means 'in mempool'), so attempting to spend a tx will prevent a tx from being re-used/expired.

niftynei commented on wallet/walletrpc.c:
```diff
@@ -1258,3 +1286,158 @@ static const struct json_command unreserveinputs_command = {
+		/* Oops we haven't reserved this utxo yet.
+		 * Let's just go ahead and reserve it now. */
+		if (utxo->status != output_state_reserved) {
+			if (utxo->status != output_state_available) {
+				return command_fail(cmd, LIGHTNINGD,
+						    "Aborting PSBT signing. UTXO %s:%u is already spent",
+						    type_to_string(tmpctx, struct bitcoin_txid,
+								   &utxo->txid),
+						    utxo->outnum);
+			}
+			/* Reserve this utxo before we spend it */
+			if (!wallet_update_output_status(cmd->ld->wallet,
+							 &txid, utxo->outnum,
+							 output_state_available,
+							 output_state_reserved))
+				return command_fail(cmd, LIGHTNINGD,
+						    "Aborting PSBT signing. Unable to reserve UTXO %s:%u",
+						    type_to_string(tmpctx, struct bitcoin_txid,
+								   &utxo->txid),
+						    utxo->outnum);
```
If someone has access to the RPC, forcing them to call 'reserveinputs' before 'signpsbt' is very weak security. :)

Let's talk a bit about *why* we do reservations in the first place. The general concept is that it stops the internal wallet from creating other transactions using that UTXO. If you attempt to use a UTXO that hasn't been marked as 'reserved', there's really no problem with this, except that there's the *possibility* another tx might be constructed that includes it.

The way it's currently written, we consider a 'sign attempt' to indicate that a tx would like to use this utxo, so if it's not already marked as reserved, we should reserve it so that no further txs constructed by the wallet will attempt to include it.

cdecker replied:
Good points. My scenario was more that we have some plugin that does some complex collaborative tx construction, and we can trust (or better we need to trust) the plugin to reserve its own UTXOs correctly, but if we just silently sign independent from our own reservation, we may receive a combined PSBT from the other participants, and pass that to `signpsbt` forgetting to check that the PSBT doesn't include our own unreserved UTXOs, thus silently attributing the UTXO to the attacker. It's not much of a safety net to require reservations, but it might just be what prevents this kind of attack.

cdecker replied:
One way to move forward here would be to split this fixup out into a separate PR and merge the rest :_)

niftynei replied:
Oh I see. One way to mitigate this would be, when we receive inputs from peers to add to the tx we're building, to verify that they're not in fact our inputs. Though this is difficult because it might be that that peer is construction a transaction with a peer who's in turn constructing a transaction with us... but in this case the UTXO would have been marked reserved by the other 'us' so it wouldn't be a problem. 

So I guess the check would be 'if this is our utxo and not already marked reserved, cancel the open with the peer', but you'd still do the check upon receiving a utxo from a peer, which, as code, doesn't exist yet.

cdecker replied:
Somehow some of my comments got lost, let me try to reconstruct them here.

My scenario was more like the following: the attacker doesn't have access to the RPC, but is negotiating some a TX that shares inputs from the victim and the attacker. This is mediated through some plugin, that takes care of the coordination, and has access to the RPC. Now the plugin does call `reserveinputs` when it collects the inputs it wants to add to the TX. The attacker pretends to do the same, but claims an input that the victim owns. If we now have the plugin call `signpsbt`, and it papers over the reservation checks, the plugin will have accidentally signed over ownership of the funds to the attacker, and it can now use those funds.

An example would be a dual-funding plugin that doesn't jump through the extra hoop of making sure that the inputs in the funding, that are claimed by the other endpoint, are not in reality its own funds. If we then sign the PSBT we happen to attribute the funds from our UTXOs, but claimed by the counterparty signed over to the initial balance of the counterparty, which is not what we want.

I think we can easily split out this and continue discussing in a separate PR, while keeping the stronger enforcement of reservations in place. That'd allow us to merge this PR, and not keep it pending.

cdecker replied:
Apparently my comments were not lost after all? GH is weird today...

cdecker replied:
> Oh I see. One way to mitigate this would be, when we receive inputs from peers to add to the tx we're building, to verify that they're not in fact our inputs. Though this is difficult because it might be that that peer is construction a transaction with a peer who's in turn constructing a transaction with us... but in this case the UTXO would have been marked reserved by the other 'us' so it wouldn't be a problem.
> 
> So I guess the check would be 'if this is our utxo and not already marked reserved, cancel the open with the peer', but you'd still do the check upon receiving a utxo from a peer, which, as code, doesn't exist yet. (and thus wouldn't impact this code patch)

Right, I'm not saying that we can't work around this issue by pushing some more validation work onto the plugin, but I'm fearful of being to lenient in what we accept, because that's how a plugin author can easily introduce a bug. It's really more of a belts-and-suspenders kind of thing here :-)

niftynei replied:
Ok. I have come around to seeing things your way. This has been removed and the tests have been updated to adjust to this requirement.

rustyrussell commented on doc/lightning-reserveinputs.7.md:
```diff
@@ -0,0 +1,79 @@
+*utxos* specifies the utxos to be used to fund the transaction, as an array
+of "txid:vout". These must be drawn from the node's available UTXO set.
+
+**reserveinputs** is similar to the first part of a **withdraw** command, but
+supports multiple outputs and uses *outputs* as parameter.
```
Probably drop this reference: withdraw will eventually become a plugin?

rustyrussell commented on doc/lightning-unreserveinputs.7.md:
```diff
@@ -0,0 +1,48 @@
+ *unreserved*, which will be true if that UTXO was successfully unreserved
+by this call.
+
+If every input in the *psbt* is unreserved by this call, *all_unreserved* will be set to true, false otherwise.
```
Don't do this, since it's trivially calculable.  Prefer minimal interfaces.

rustyrussell commented on wallet/walletrpc.c:
```diff
@@ -1258,3 +1286,158 @@ static const struct json_command unreserveinputs_command = {
+		      tal_hex(tmpctx, msg));
+
+	response = json_stream_success(cmd);
+	json_add_psbt(response, "signed_psbt", signed_psbt);
```
Maybe add  abool indiciating if it's ready for sendpsbt? (i.e. finalizable)

niftynei replied:
will do later.

rustyrussell commented on bitcoin/tx.c:
```diff
@@ -114,11 +114,16 @@ struct amount_sat bitcoin_tx_compute_fee_w_inputs(const struct bitcoin_tx *tx,
+			return AMOUNT_SAT(0);
+
 		ok = amount_sat_sub(&input_val, input_val,
 				    amount_asset_to_sat(&asset));
```
This is equivalent to:
```
if (!ok)
    return AMOUNT_SAT(0);
```

niftynei replied:
oh nice point. fixed.

rustyrussell commented:
Ack, but needs rebase:

1. Empty whitespace-only commit?
2. Commit already in master which just adds a test?

But some of this may be GH weirdness...

rustyrussell commented:
> Quite a nice change, making the internal wallet a lot more like `bitcoind`'s. Can we deprecate `txprepare` and build on this instead?

We should do that once we have dual funding, because I'd rather not change interfaces twice...

niftynei commented:
>Ack, but needs rebase:

> Empty whitespace-only commit?
> Commit already in master which just adds a test?

Ok, updated requested changes and removed/fixed the two commit weirdness that happened due to master rebase. (one removed, the other folded into the appropriate commit).

cdecker commented:
Looks like all the requested changes were implemented, except the cosmetic change to add fully signed flag in the `signpsbt` result. I'll nudge Travis to completion and merge asap.

ACK https://github.com/ElementsProject/lightning/pull/3775/commits/e101bc88f539bfcfc6fe2d428002bbe4a822f673
